packages/component/src/providers/CustomElements/customElements/registerCodeBlockCopyButton.ts

```diff
@@ -0,0 +1,125 @@
+    buttonElement.dataset.testid = testIds.codeBlockCopyButton;
+    buttonElement.type = 'button';
+
+    buttonElement.append(copyIconImageElement);
+    buttonElement.append(copiedIconImageElement);
+
+    this.append(buttonElement);
+
+    const showAsPressed = () => {
+      buttonElement.ariaPressed = 'true';
+      buttonElement.classList.add('webchat__code-block-copy-button--copied');
+      copiedIconImageElement.ariaHidden = undefined;
+      copyIconImageElement.ariaHidden = 'true';
+    };
+
+    const showAsUnpressed = () => {
+      buttonElement.ariaPressed = undefined;
+      buttonElement.classList.remove('webchat__code-block-copy-button--copied');
+      copiedIconImageElement.ariaHidden = 'true';
+      copyIconImageElement.ariaHidden = undefined;
+    };
+
+    // Initially, show as unpressed.
+    showAsUnpressed();
+
+    buttonElement.addEventListener('animationend', () => showAsUnpressed());
+
+    buttonElement.addEventListener('click', () => {
+      if (buttonElement.ariaDisabled === 'true') {
+        return;
+      }
+
+      (async () => {
+        try {
+          const { state } = await navigator.permissions.query({ name: 'clipboard-write' as any });
+
+          if (state === 'granted') {
+            await navigator.clipboard?.write([
+              new ClipboardItem({ 'text/plain': new Blob([this.dataset.value], { type: 'text/plain' }) })
+            ]);
+
+            showAsPressed();
+          } else if (state === 'denied') {
+            buttonElement.ariaDisabled = 'true';
+          }
+        } catch (error) {
+          console.warn('botframework-webchat: Failed to copy code block to clipboard.', error);
+        }
+      })();
+    });
+
+    this.#buttonElement = buttonElement;
+    this.#copiedIconImageElement = copiedIconImageElement;
+    this.#copyIconImageElement = copyIconImageElement;
+  }
+
+  #buttonElement: HTMLButtonElement;
+  #copiedIconImageElement: HTMLDivElement;
+  #copyIconImageElement: HTMLDivElement;
+
+  attributeChangedCallback(name: ObservedAttributes, oldValue: string, newValue: string) {
+    if (name === 'class') {
+      this.#buttonElement.classList.remove(oldValue);
+      this.#buttonElement.classList.add(newValue);
+    } else if (name === 'data-alt-copied') {
+      this.#copiedIconImageElement.ariaLabel = this.dataset.altCopied;
+    } else if (name === 'data-alt-copy') {
+      this.#copyIconImageElement.ariaLabel = this.dataset.altCopy;
+    }
+  }
+}
+
+export default function registerCodeBlockCopyButton(hash: string): string {
+  // Allowed tag names are specified here, https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define#valid_custom_element_names
+  const name = `webchat-${hash}--code-block-copy-button`;
+
+  customElements.define(
+    name,
+    // Using anonymous class because one class can only be registered with one tag name.
+    // Web Components enforce 1:1 relationship.
+    class extends CodeBlockCopyButtonElement {
```

Comment:
If you don't want to create a chain of inheritance, it should be possible to wrap the class creation into a function and call here.

On the other hand this approach allows ckecking if we derived from the initial class.

Overall, there shouldn't be a problem keeping 1-1 relationship between Web Chat instances using React primitives. I assume something else needs our attention.